src/qibolab/instruments/qm.py

```diff
@@ -597,28 +620,59 @@ def declare_output(self, options, threshold=None, angle=None):
             raise_error(ValueError, f"Invalid acquisition type {acquisition_type}.")
         self.acquisition.assign_element(self.element)
 
-    def bake(self, config: QMConfig):
-        if self.baked is not None:
-            raise_error(RuntimeError, f"Bake was already called for {self.pulse}.")
-        # ! Only works for flux pulses that have zero Q waveform
-
-        # Create the different baked sequences, each one corresponding to a different truncated duration
-        # for t in range(self.pulse.duration + 1):
-        with baking(config.__dict__, padding_method="right") as self.baked:
-            # if t == 0:  # Otherwise, the baking will be empty and will not be created
-            #    wf = [0.0] * 16
-            # else:
-            # wf = waveform[:t].tolist()
-            if self.pulse.duration == 0:
-                waveform = [0.0] * 16
-            else:
-                waveform = self.pulse.envelope_waveform_i.data.tolist()
-            self.baked.add_op(self.pulse.serial, self.element, waveform)
-            self.baked.play(self.pulse.serial, self.element)
-            # Append the baking object in the list to call it from the QUA program
-            # self.segments.append(b)
 
-        self.duration = self.baked.get_op_length()
+DurationsType = Union[List[int], npt.NDArray[int]]
+
+
+@dataclass
+class BakedPulse(QMPulse):
+    """Baking allows 1ns resolution in the pulse waveforms."""
+
+    segments: List[Baking] = field(default_factory=list)
+    """Baked segments implementing the pulse."""
+    amplitude: Optional[float] = None
+    """Amplitude of the baked pulse. Relevant only when sweeping amplitude."""
+    durations: Optional[DurationsType] = None
+
+    def __hash__(self):
+        return super().__hash__()
+
+    @property
+    def duration(self):
+        return self.segments[-1].get_op_length()
+
+    def bake(self, config: QMConfig, durations: DurationsType):
+        self.segments = []
+        self.durations = durations
+        for t in durations:
+            with baking(config.__dict__, padding_method="right") as segment:
+                if t == 0:  # Otherwise, the baking will be empty and will not be created
+                    waveform = [0.0] * 16
+                else:
+                    original_waveform = self.pulse.envelope_waveform_i.data.tolist()
+                    expanded_waveform = list(original_waveform)
```

Comment:
Will this work for DRAG pulses that need both i and q ?

Reply:
No, currently the baked pulse has only one i-component, as it is only used for flux pulses which are sent through a single port. It is possible and easy to [bake both i and q](https://github.com/qua-platform/py-qua-tools/blob/main/qualang_tools/bakery/README.md#how-can-i-add-operations-inside-the-baking-context-manager), I am just not sure if it is needed, at least for now, as drive and readout pulses are never baked.